unittests/comprehensive/test_ForceDependentSlip.cpp

```diff
@@ -0,0 +1,280 @@
+
+  const double slip2 = 0.03;
+  // Test slip compliance in the secondary friction direction
+  body1Dynamics->setSlipCompliance(0);
+  body1Dynamics->setSecondarySlipCompliance(slip2);
+
+  EXPECT_DOUBLE_EQ(body1Dynamics->getSlipCompliance(), 0.0);
+  EXPECT_DOUBLE_EQ(body1Dynamics->getSecondarySlipCompliance(), slip2);
+
+  // Step without external force so the body stop moving
+  for (auto i = 0u; i < 500; ++i)
+  {
+    world->step();
+  }
+  EXPECT_NEAR(0.0, body1->getLinearVelocity().x(), 1e-4);
+  EXPECT_NEAR(0.0, body1->getLinearVelocity().y(), 1e-4);
+
+  // The second box should remain at rest because of friction
+  EXPECT_NEAR(0.0, body2->getLinearVelocity().x(), 1e-4);
+  EXPECT_NEAR(0.0, body2->getLinearVelocity().y(), 1e-4);
+
+  // Apply force in the +y direction
+  for (auto i = 0u; i < numSteps; ++i)
+  {
+    body1->addExtForce({0, extForce, 0});
+    body2->addExtForce({0, extForce, 0});
+    world->step();
+
+    if (i > 1500)
+    {
+      // The velocity of body1 should stabilize at F_ext * slip2 = 0.3 m/s
+      EXPECT_NEAR(0.0, body1->getLinearVelocity().x(), 1e-4);
+      EXPECT_NEAR(extForce * slip2, body1->getLinearVelocity().y(), 1e-4);
+
+      // The second box should remain at rest because of friction
+      EXPECT_NEAR(0.0, body2->getLinearVelocity().x(), 1e-4);
+      EXPECT_NEAR(0.0, body2->getLinearVelocity().y(), 1e-4);
+    }
+  }
+}
+
+//==============================================================================
+// Test two cylinders, one with its z axis pointing in the z axis of the world
+// so it's purely slipping, and the other with its z axis pointing in the y axis
+// of the world so it's rolling and slipping.
+TEST(ForceDependentSlip, CylinderSlipVelocity)
+{
+  using Eigen::Vector3d;
+  const double mass = 2.0;
+  const double radius = 0.5;
+  const double slip = 0.02;
+  auto skeleton1 = createCylinder(radius, 0.3, {-5, 0, radius});
+  skeleton1->setName("Skeleton1");
+
+  auto skeleton2 = createCylinder(
+      radius, 0.8, {5.0, 0, radius}, {math::constantsd::half_pi(), 0, 0});
+  skeleton2->setName("Skeleton2");
+
+  auto body1 = skeleton1->getRootBodyNode();
+  body1->setMass(5.0);
+  auto body1Dynamics = body1->getShapeNode(0)->getDynamicsAspect();
+
+  EXPECT_DOUBLE_EQ(body1Dynamics->getFrictionCoeff(), 1.0);
+
+  body1Dynamics->setSlipCompliance(slip);
+  body1Dynamics->setFirstFrictionDirection(Vector3d::UnitX());
+  EXPECT_DOUBLE_EQ(body1Dynamics->getSlipCompliance(), slip);
+  EXPECT_EQ(body1Dynamics->getFirstFrictionDirection(), Vector3d::UnitX());
+
+  auto body2 = skeleton2->getRootBodyNode();
+  body2->setMass(mass);
+  auto body2Dynamics = body2->getShapeNode(0)->getDynamicsAspect();
+  EXPECT_DOUBLE_EQ(body2Dynamics->getFrictionCoeff(), 1.0);
+
+  // Set the friction direction to +z in the shape frame because it always be
+  // orthogonal to the floor's normal.
+  body2Dynamics->setFirstFrictionDirection(Vector3d::UnitZ());
+  body2Dynamics->setSecondarySlipCompliance(slip);
```

Comment:
so let me see if I have this straight: the first friction direction is in the body-fixed Z direction, which is the world Y direction. and you are setting the non-zero slip value on this body in the secondary slip direction, which must be the world X direction? I think it would be helpful to spell this out

Reply:
I set them farther apart in their y coordinates. 334a937